connection/mysql.go

```diff
@@ -702,26 +702,51 @@ func (dbw *DBWrapper) SqlBulkInsert(rows []Row, stmt *BulkInsertStmt) error {
 
 	DbBulkInserts.Inc()
 
-	placeholders := make([]string, len(rows))
-	values := make([]interface{}, len(rows)*stmt.NumField)
-	j := 0
+	done := make(chan struct{})
+	defer close(done)
 
-	for i, r := range rows {
-		placeholders[i] = stmt.Placeholder
+	chErr := make(chan error)
+	wg := sync.WaitGroup{}
+	wgDone := make(chan struct{})
 
-		for _, v := range r.InsertValues() {
-			values[j] = v
-			j++
-		}
-	}
+	for b := range ChunkRows(done, rows, 500) {
+		wg.Add(1)
+		go func(bulk []Row) {
+			defer wg.Done()
+			placeholders := make([]string, len(bulk))
+			values := make([]interface{}, len(bulk)*stmt.NumField)
+			j := 0
 
-	query := fmt.Sprintf(stmt.Format, strings.Join(placeholders, ", "))
+			for i, r := range bulk {
+				placeholders[i] = stmt.Placeholder
 
-	_, err := dbw.WithRetry(func() (result sql.Result, e error) {
-		return dbw.SqlExec(mysqlObservers.bulkInsert, query, values...)
-	})
+				for _, v := range r.InsertValues() {
+					values[j] = v
+					j++
+				}
+			}
 
-	if err != nil {
+			query := fmt.Sprintf(stmt.Format, strings.Join(placeholders, ", "))
+
+			_, err := dbw.WithRetry(func() (result sql.Result, e error) {
+				return dbw.SqlExec(mysqlObservers.bulkInsert, query, values...)
+			})
+
+			if err != nil {
+				chErr <- err
```

Comment:
If both of two operations fail, one writes to chErr successfully as you read from it in the select below. The other one will deadlock.

Reply:
Whoops, that's bad... Shouldn't have overlooked that. Thanks!